language-plutus-core/src/Language/PlutusCore/Instance/Eq/Internal.hs

```diff
@@ -0,0 +1,263 @@
+{-# LANGUAGE RankNTypes            #-}
+{-# LANGUAGE TemplateHaskell       #-}
+{-# LANGUAGE TypeApplications      #-}
+{-# LANGUAGE UndecidableInstances  #-}
+
+module Language.PlutusCore.Instance.Eq.Internal
+    ( LR (..)
+    , RL (..)
+    , EqRename
+    , ScopedEqRename
+    , runEqRename
+    , withTwinBindings
+    , eqNameM
+    , eqTypeM
+    , eqTermM
+    , eqProgramM
+    ) where
+
+import           PlutusPrelude
+
+import           Language.PlutusCore.Name
+import           Language.PlutusCore.Rename.Monad
+import           Language.PlutusCore.Type
+
+import           Control.Lens
+
+-- | From left to right.
+newtype LR a = LR
+    { unLR :: a
+    } deriving (Generic)
+
+-- | From right to left.
+newtype RL a = RL
+    { unRL :: a
+    } deriving (Generic)
+
+-- | A left @a@ and a right @a@.
+data Bilateral a = Bilateral
+    { _bilateralL :: a
+    , _bilateralR :: a
+    }
+
+makeLenses ''Bilateral
+
+instance Wrapped (LR a)
+instance Wrapped (RL a)
+instance HasUnique name unique => HasUnique (LR name) (LR unique)
+instance HasUnique name unique => HasUnique (RL name) (RL unique)
+
+instance Semigroup a => Semigroup (Bilateral a) where
+    Bilateral l1 r1 <> Bilateral l2 r2 = Bilateral (l1 <> l2) (r1 <> r2)
+
+instance Monoid a => Monoid (Bilateral a) where
+    mempty = Bilateral mempty mempty
+
+-- To rename from left to right is to update the left renaming.
+instance HasRenaming ren unique => HasRenaming (Bilateral ren) (LR unique) where
+    renaming = bilateralL . renaming . coerced @(Renaming unique)
+
+-- To rename from right to left is to update the right renaming.
+instance HasRenaming ren unique => HasRenaming (Bilateral ren) (RL unique) where
+    renaming = bilateralR . renaming . coerced @(Renaming unique)
+
+-- | The type of a runnable equality check. @Maybe ()@ is isomorphic to 'Bool' and we use it
+-- instead of 'Bool', because this unlocks the convenient and readable do-notation and allows for
+-- automatic short-circuiting, which would be tedious with @Rename (Bilateral ren) Bool@.
+type EqRename ren = RenameT (Bilateral ren) Maybe ()
+type ScopedEqRename = EqRename ScopedRenaming
+
+-- | Run an 'EqRename' computation.
+runEqRename :: Monoid ren => EqRename ren -> Bool
+runEqRename = isJust . runRenameT
+
+{- Note [Modulo alpha]
+The implemented algorithm of checking equality modulo alpha works as follows
+(taking types as an example):
+
+- traverse both the types simultaneously
+- if the outermost constructors differ, return 'False'
+- otherwise if the constructors are binders, then record that the two possibly distinct bound
+    variables map to each other
+- otherwise if the constructors are variables, look them up in the current scope
+    * if both are in scope, then those are bound variables, so check that they map to each other
+        (i.e. are introduced by twin binders)
+    * if both are not in scope, then those are free variables, so check their direct equality
```

Comment:
> direct equality

That would be name equality modulo alpha (aka stable under renames)?

Reply:
Modulo alpha only applies for things with binders, but here we just check equality of two variables. Changed the wording to "if both are not in scope, then those are free variables, so check if their uniques are equal".